controller/src/main/java/com/epam/esm/controller/exception/ApplicationExceptionHandler.java

```diff
@@ -0,0 +1,84 @@
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.springframework.context.support.ReloadableResourceBundleMessageSource;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import com.epam.esm.service.exception.CustomErrorCode;
+import com.epam.esm.service.exception.ExceptionDetails;
+import com.epam.esm.service.exception.GiftCertificateDtoValidationException;
+import com.epam.esm.service.exception.NoSuchResourceException;
+import com.epam.esm.service.exception.PageException;
+import com.epam.esm.service.exception.TagValidationException;
+
+@ControllerAdvice
+public class ApplicationExceptionHandler {
+
+    public static final String NO_SUCH_RESOURCE_MESSAGE = "no_resource";
+
+    private final ReloadableResourceBundleMessageSource resourceBundle;
+
+    public ApplicationExceptionHandler(ReloadableResourceBundleMessageSource resourceBundle) {
+        this.resourceBundle = resourceBundle;
+    }
+
+    @ExceptionHandler(NoSuchResourceException.class)
+    public ResponseEntity<ExceptionDetails> handleNoSuchResourceException(NoSuchResourceException exception) {
+        HttpStatus status = HttpStatus.NOT_FOUND;
+        String message = getErrorResponse(NO_SUCH_RESOURCE_MESSAGE);
+        String errorCode = status.value() + exception.getCode();
+        ExceptionDetails data = new ExceptionDetails(LocalDateTime.now(), status.value(), message, errorCode);
+
+        return new ResponseEntity<>(data, status);
+    }
+
+    @ExceptionHandler(TagValidationException.class)
+    public ResponseEntity<ExceptionDetails> handleTagValidationException(TagValidationException exception) {
+        HttpStatus status = HttpStatus.BAD_REQUEST;
+        String message = getErrorResponse(exception.getMessage());
+        String exceptionCode = CustomErrorCode.TAG.getCode();
+        String errorCode = status.value() + exceptionCode;
+        ExceptionDetails data = new ExceptionDetails(LocalDateTime.now(), status.value(), message, errorCode);
+        return new ResponseEntity<>(data, status);
+    }
+
+    @ExceptionHandler(GiftCertificateDtoValidationException.class)
+    public ResponseEntity<ExceptionDetails> handleGiftCertificateDtoValidationException(
+            GiftCertificateDtoValidationException exception) {
+        HttpStatus status = HttpStatus.BAD_REQUEST;
+        String message = getErrorResponse(exception.getMessage());
+        String exceptionCode = CustomErrorCode.CERTIFICATE.getCode();
+        String errorCode = status.value() + exceptionCode;
+        ExceptionDetails data = new ExceptionDetails(LocalDateTime.now(), status.value(), message, errorCode);
+        return new ResponseEntity<>(data, status);
```

Comment:
Seems like this part can be extracted to separate method.

Reply:
Separate createResponseEntity() was created, handleTagValidationException(TagValidationException exception), handleGiftCertificateDtoValidationException(GiftCertificateDtoValidationException exception), handlePageException(PageException exception) use it